connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java

```diff
@@ -124,6 +124,9 @@ public String toString() {
     public static final int NO_PARTITIONS = -1;
     public static final short NO_REPLICATION_FACTOR = -1;
 
+    private static final String DEFAULT_ADMIN_CLIENT_RETRIES = "10";
+    private static final String DEFAULT_ADMIN_CLIENT_BACKOFF_MS = "100";
+
```

Comment:
Okay, so I was verifying what these defaults really were on the `AdminClientConfig`, and I noticed that [AdminClientConfig.RETRIES_CONFIG](https://github.com/apache/kafka/blob/2ccc834faa3fffcd5d15d2463aeef3ee6f5cea13/clients/src/main/java/org/apache/kafka/clients/admin/AdminClientConfig.java#L173-L178) defaults `Int.MAX_VALUE`. I did a bit of research, and found [KIP-533](https://cwiki.apache.org/confluence/display/KAFKA/KIP-533%3A+Add+default+api+timeout+to+AdminClient) changed the defaults, and I don't know that we can/should count on `AdminClientConfig.RETRIES`.

I consulted with @kkonstantine, and after a bit of discussion we agree that it would be better to not make these configurable and to instead define the maximum duration to try to read the offsets as a constant (e.g., 15 minutes), plus a constant for retry backoff millis (e.g., 10 seconds). The max duration is easier to reason about.

The 15 minutes is a balance between retrying long enough to work with most incidents of metadata propagation taking too long, and short enough to fail if the worker cannot start up.

Interestingly, I don't think the duration needs to be defined in `TopicAdmin`, and is better defined in `KafkaBasedLog.start()`. That way, the `KafkaBasedLog` passes the duration into the `TopicAdmin.endOffset(Set<TopicPartition> partitions, Duration timeout)` method, which is the method that calls the `RetryUtil.retry(Callable<T> callable, Duration retryTimeout, long retryBackoffMs)`.

This isn't too much different than what you have, but I do think it's cleaner. I don't think we have seen too many retriable exception during `KafkaBasedLog.start()`, which means it's kind of an edge case.

WDYT?



Second, it might be good to add these methods:
```
public Map<TopicPartition, Long> retryEndOffsets(Set<TopicPartition> partitions, Duration timeout) { ... }
```
that retries for the given timeout while using `DEFAULT_ADMIN_CLIENT_BACKOFF_MS` for backoffs. Then, if you change `KafkaBasedLog.readEndOffsets(...)` and `e the `KafkaBasedLog.readToLogEnd(...)` to take a `Duration timeout` rather than a boolean, then `KafkaBasedLog.start() could just call `readToLogEnd(Duration.ofMinutes(15))` while `KafkaBasedLog.WorkThread.run()` could call `readToLogEnd(null)`.

You'd need to refactor your `retry(...)` method a bit to take the `Duration`. You could also